src/main/java/org/araymond/joal/core/ttorrent/client/Client.java

```diff
@@ -61,161 +71,169 @@ void setAnnouncerExecutor(final AnnouncerExecutor announcerExecutor) {
+        Lock lock = this.lock.writeLock();
         try {
-            this.lock.writeLock().lock();
-            this.currentlySeedingAnnouncer.remove(announcer); // Remove from announcers list asap, otherwise the deletion will trigger a announce stop event.
-            this.torrentFileProvider.moveToArchiveFolder(announcer.getTorrentInfoHash());
-            this.addTorrent();
+            lock.lock();
+            this.currentlySeedingAnnouncers.remove(announcer);  // Remove from announcers list asap, otherwise the deletion will trigger an announce stop event.
+            if (!this.stop) {
+                this.torrentFileProvider.moveToArchiveFolder(announcer.getTorrentInfoHash());
+                this.addTorrentFromDirectory();
+            }
         } catch (final NoMoreTorrentsFileAvailableException ignored) {
         } finally {
-            this.lock.writeLock().unlock();
+            lock.unlock();
         }
     }
 
     public void onNoMorePeers(final InfoHash infoHash) {
-        if (!this.appConfiguration.isKeepTorrentWithZeroLeechers()) {
+        if (!this.appConfig.isKeepTorrentWithZeroLeechers()) {
             this.torrentFileProvider.moveToArchiveFolder(infoHash);
         }
     }
 
     public void onTorrentHasStopped(final Announcer stoppedAnnouncer) {
-        if (this.stop) {
-            this.currentlySeedingAnnouncer.remove(stoppedAnnouncer);
-            return;
-        }
+        Lock lock = this.lock.writeLock();
         try {
-            this.lock.writeLock().lock();
-
-            this.addTorrent();
+            lock.lock();
+            if (!this.stop) {
+                this.addTorrentFromDirectory();
+            }
         } catch (final NoMoreTorrentsFileAvailableException ignored) {
         } finally {
-            this.currentlySeedingAnnouncer.remove(stoppedAnnouncer);
-            this.lock.writeLock().unlock();
+            this.currentlySeedingAnnouncers.remove(stoppedAnnouncer);
+            lock.unlock();
         }
     }
 
     @Override
     public void onTorrentFileAdded(final MockedTorrent torrent) {
         this.eventPublisher.publishEvent(new TorrentFileAddedEvent(torrent));
-        if (this.stop) {
-            return;
-        }
-        try {
-            this.lock.writeLock().lock();
-            if (this.currentlySeedingAnnouncer.size() >= this.appConfiguration.getSimultaneousSeed()) {
-                return;
+        if (!this.stop && this.currentlySeedingAnnouncers.size() < this.appConfig.getSimultaneousSeed()) {
+            Lock lock = this.lock.writeLock();
+            try {
+                lock.lock();
+                addTorrent(torrent);
+            } finally {
+                lock.unlock();
             }
-            final Announcer announcer = this.announcerFactory.create(torrent);
-            this.currentlySeedingAnnouncer.add(announcer);
-            this.delayQueue.addOrReplace(AnnounceRequest.createStart(announcer), 1, ChronoUnit.SECONDS);
```

Comment:
Note instead of this duplication, we call the newly introduced `addTorrent(torrent)` method.
The new method uses `interval` of 0 instead of 1 as was before. Hope this isn't going to cause any issues?